### PR TITLE
Document limitations of Bounter

### LINKED ISSUE #36: Include limitations of bounter in ReadMe for non-power users.
As per the discussion [here]( https://news.ycombinator.com/item?id=15521177), I think it would be good to include a section like:

## Where bounter fails/is inaccurate/Accuracy TradeOffs
I refer to the part below:
>The introduction is misleading about what tradeoffs it makes
  >>> from bounter import bounter
  >>> bounts = bounter(size_mb=1)
  >>> bounts.update(str(i) for i in xrange(1000000))
  >>> bounts['100']
  0L
should give 1. The description should be more upfront that you are getting inaccurate answers.
> Bounter implements approximative algorithms using optimized low-level C structures, to avoid the overhead of Python objects.

"approximative algorithms" isn't a commonly used phrase (at least according to Google) and this sentence doesn't say anything about what the actual trade-off is (loss of accuracy in the counts). A possible alternative would be writing to hard disk so what trade-off is made isn't obvious.

I also don't know what the precision and recall percentages in the example table mean
It will be good for new comers/non-power users.

Wdyt @menshikh-iv ?

I would be glad to make a PR for this.

### PR DESCRIPTION
**Pull request by aneesh-joshi**
Credit goes to the people on [this](https://news.ycombinator.com/item?id=15521177) page.
I merely worded their concerns.

Fixes #36 .

CC: @menshikh-iv

### CODE REVIEW

**Comment by menshikh-iv** [comment]
Thanks @aneesh-joshi 

@piskvorky any comments?

File: README.md [hunk 1]
```diff
@@ -26,7 +26,22 @@ However, unlike `dict` or `Counter`, Bounter can process huge collections where
...
+
+
+## When not to use Bounter?
+Beware, Bounter is only a probabilistic frequency counter and cannot be relied on for fine counting. (You can't expect a data structure with finite size to hold infinite data.)
+Example of Bounter failing:
+
+```python
+from bounter import bounter
+bounts = bounter(size_mb=1)
+bounts.update(str(i) for i in range(1000000))
```
Comment (piskvorky) [inline README.md]: Add another 0 (`10000000`), to make it clear this is not related to 1 MB.

File: README.md [hunk 2]
```diff
@@ -26,7 +26,22 @@ However, unlike `dict` or `Counter`, Bounter can process huge collections where
 
 Bounter implements approximative algorithms using optimized low-level C structures, to avoid the overhead of Python objects. It lets you specify the maximum amount of RAM you want to use. In the Wikipedia example below, Bounter uses 31x less memory compared to `Counter`.
 
-Bounter is also marginally faster than the built-in `dict` and `Counter`, so wherever you can represent your **items as strings** (both byte-strings and unicode are fine, and Bounter works in both Python2 and Python3), there's no reason not to use Bounter instead.
+Bounter is also marginally faster than the built-in `dict` and `Counter`, so wherever you can represent your **items as strings** (both byte-strings and unicode are fine, and Bounter works in both Python2 and Python3), there's no reason not to use Bounter instead except:
+
+
+## When not to use Bounter?
+Beware, Bounter is only a probabilistic frequency counter and cannot be relied on for fine counting. (You can't expect a data structure with finite size to hold infinite data.)
```
Comment (piskvorky) [inline README.md]: `fine` => `exact`

File: README.md [hunk 3]
```diff
@@ -26,7 +26,22 @@ However, unlike `dict` or `Counter`, Bounter can process huge collections where
...
+
+```python
+from bounter import bounter
+bounts = bounter(size_mb=1)
+bounts.update(str(i) for i in range(1000000))
+bounts['100']
+0
+```
+
+Please use `Counter` or `dict` when such fine counts matter. When they don't matter, like in most NLP applications with a huge corpora, Bounter is a very good alternative.
```
Comment (piskvorky) [inline README.md]: `fine` => `exact`.

Comment (piskvorky) [inline README.md]: `…like in most NLP and ML applications with huge datasets, …`

File: README.rst [hunk 4]
```diff
@@ -38,7 +38,23 @@ below, Bounter uses 31x less memory compared to ``Counter``.
 Bounter is also marginally faster than the built-in ``dict`` and
 ``Counter``, so wherever you can represent your **items as strings**
 (both byte-strings and unicode are fine, and Bounter works in both
-Python2 and Python3), there's no reason not to use Bounter instead.
+Python2 and Python3), there's no reason not to use Bounter instead except:
+
+When not to use Bounter?
+------------------------
+
+Beware, Bounter is only a probabilistic frequency counter and cannot be relied on for fine counting. (You can't expect a data structure with finite size to hold infinite data.)
```
Comment (piskvorky) [inline README.rst]: Is this file auto-generated? Or why are we keeping two parallel READMEs?

Anyway, the same changes here please.

**Comment by piskvorky** [comment]
Good idea! Thanks.

File: README.rst (hunk 4 above)
Comment (menshikh-iv) [reply README.rst]: In this case, I propose to stay only `.rst` (because this used for PyPI too) and drop markdown version (to don't support 2 almost same versions)

Comment (piskvorky) [reply README.rst]: I wonder if they're out of sync already… @aneesh-joshi can you check?

Comment (aneesh-joshi) [reply README.rst]: @piskvorky 
I will check. The .rst is not auto generated. I manually edited it.

**Comment by menshikh-iv** [comment]
@piskvorky README rst & md really not synced yet. I still think that we should stay only one version (create best .rst version & drop markdown version to avoid "de-sync" state).

**Comment by piskvorky** [comment]
> I still think…

Did we already discuss this? To me a single version (`.rst`) sounds better too, but I don't remember the pros/cons. What were the arguments against it? Why are we keeping both `.rst` and `.md`?

**Comment by menshikh-iv** [comment]
I don't remember (I'm even not sure if we discussed this)

Maybe @isamaru remember? 

Right now I see no pros to maintain both `.md` and `.rst`, I prefer `.md`, but PyPI can render only `.rst`, for this reason, single `.rst` should be our choice.

**Comment by aneesh-joshi** [comment]
@menshikh-iv @piskvorky 
I've made the mentioned changes.
Waiting on your decision on whether to keep .rst or .md

Some questions:
- Does .rst render properly on github?
- How do I autogenerate the .rst? (I can't remember it and couldn't get the exact one through google)

File: README.md [hunk 5]
```diff
@@ -30,18 +30,18 @@ Bounter is also marginally faster than the built-in `dict` and `Counter`, so whe
...
 from bounter import bounter
 bounts = bounter(size_mb=1)
-bounts.update(str(i) for i in range(1000000))
+bounts.update(str(i) for i in range(10000000))
 bounts['100']
 0
 ```
 
-Please use `Counter` or `dict` when such fine counts matter. When they don't matter, like in most NLP applications with a huge corpora, Bounter is a very good alternative.
+Please use `Counter` or `dict` when such exact counts matter. When they don't matter, like in most NLP and ML applications with a huge datasets, Bounter is a very good alternative.
```
Comment (piskvorky) [inline README.md]: `a huge datasets` => `huge datasets`.

File: README.rst [hunk 6]
```diff
@@ -43,18 +43,18 @@ Python2 and Python3), there's no reason not to use Bounter instead except:
...
 
     from bounter import bounter
     bounts = bounter(size_mb=1)
-    bounts.update(str(i) for i in range(1000000))
+    bounts.update(str(i) for i in range(10000000))
     bounts['100']
     0
 
-Please use ``Counter`` or ``dict`` when such fine counts matter. When they don't matter, like in most NLP applications with a huge corpora, Bounter is a very good alternative.
+Please use ``Counter`` or ``dict`` when such exact counts matter. When they don't matter, like in most NLP and ML applications with a huge datasets, Bounter is a very good alternative.
```
Comment (piskvorky) [inline README.rst]: dtto

File: README.md (hunk 5 above)
Comment (aneesh-joshi) [reply README.md]: Shoot! Missed that somehow.

**Comment by aneesh-joshi** [comment]
@menshikh-iv 
I autogenerated the .rst with
`pandoc --from=markdown --to=rst --output=README.rst README.md`

but the generated file has a huge diff compared with the original .rst

**Comment by menshikh-iv** [comment]
@aneesh-joshi that's expected: if you looking into commit history, you'll see a huge diff between this 2 files. 

Here you need to do some manual work ("join" both to one file).

**Comment by aneesh-joshi** [comment]
@menshikh-iv 
If you [see the diff](https://github.com/RaRe-Technologies/bounter/pull/37/files), there is a very large change. Editing it manually will be tedious.
Is there a version mismatch or config mismatch in the pandocs?
Can i just keep the[ old version](https://github.com/RaRe-Technologies/bounter/blob/f8c84bbe865ef0b9f8a2fc7676bb2ee02f435bc1/README.rst) of the rst?

**Comment by menshikh-iv** [comment]
@aneesh-joshi okay, but please raise an issue about syncing `.rst` and `.md`

**Comment by aneesh-joshi** [comment]
@menshikh-iv 
No problem. I fixed it (I think?).

File: README.rst [hunk 7]
```diff
@@ -272,5 +270,5 @@ Technologies <https://rare-technologies.com/>`__
    :target: https://groups.google.com/forum/#!forum/gensim
 .. |Gitter| image:: https://img.shields.io/badge/gitter-join%20chat%20%E2%86%92-09a3d5.svg
    :target: https://gitter.im/RaRe-Technologies/gensim
-.. |Follow| image:: https://img.shields.io/twitter/follow/gensim_py.svg?style=social&label=Follow
+.. |Follow| image:: https://img.shields.io/twitter/follow/spacy_io.svg?style=social&label=Follow
```
[the remark this answers is not in the document]
Comment (aneesh-joshi) [reply README.rst]: @piskvorky I think you're looking at an outdated diff. Although I can't understand why `pandoc` replaced gensim_py with spacy_io and how it did it.

@menshikh-iv Is this merge worthy?

[the remark this answers is not in the document]
Comment (piskvorky) [reply README.rst]: No, I mean the current version still contains the spacy badge. You fixed it previously -- please use that version. Thanks!

**Comment by aneesh-joshi** [comment]
ping @menshikh-iv @piskvorky

**Comment by menshikh-iv** [comment]
Thanks @aneesh-joshi :rocket:
